### DISCUSSION TITLE
Broken USB on MATEK 722 STD : any alternative way to connect by UART ? - Lot of confusion about

### DISCUSSION THREAD

**Discussion by alccrl** [iNavFlight/inav#8400] [Q&A]
Hi.

On Internet there is much confusion about solutions to have a second connection to PC , other then USB MSP/VCP.
Online Videos seem to be not updated with last Inav versions or fc hw, and most refer Betaflight, not Inav.

On Matek F722 STD FC I tried FTDI, direct USB2Serial, DFU on UART1, but no way to communicate between FC and inav-configurator.
Serial ports (usb2serial or ftdi) are correctly recognized by PC, ST programmer utility (set to UART) complains no connection when FC is booted in DFU mode. 

It is not even clear how to feed vcc to the system, 3.3v or 5v.
Direct usb2serial protocol and hw sufficies three wires (GND, TX, RX) if FC and PC are already powered, whilst FTDI may convert USB from PC to 5v or 3.3v and is put in the middle between pc and fc.
In the Mateksys faqs there is a note suggesting connections but their solution does not work, and it is quite an obsolete post. 

What has to be done in the inav configurator for enabling secondary connection to pc ?
Is it necessary to use a free uart (I have UART5 unused) or just solder tx1 rx1 (Uart1) port ?

Does anybody have any idea ?

Since DFU mode does not work on TX1/RX1, I guess that port 1 cannot be used for this purpose on this FC, but this is against declaration on mateksys site that assert  uart1 can be used for USB connection alternative.

So,
combinations are so many, either sw or hw, and I would like to keep my fc, not to change it because of a stupid usb dying connector ...!

My case:
fc : matek f722 std
fw : inav 5.1
FTDI FT232RL USB to TTL Serial Converter Adapter Module 5V and 3.3V
Serial2USB commercial cable with 9 pin connector
pc : laptop with both Win10 and Linux
USB cable with microUSB fc side
USB cable with miniUSB connector to be used on FTDI side.
FTDI board connected to UART1 on fc side and miniUSB connector pc side

Thanks in advance,

Alfredo

**Comment by stronnag** [comment]
It is certainly possible to flash MATEK baords using UART1. I have a F405 with broken USB connector, which I regularly flash development builds via UART1.

* It is **not** in DFU mode, it uses the ST serial bootloader  (DFU requires **USB**).
* You should use `stm32flash` or similar tool that uses the serial bootloader.

For completeness ($ represents the shell prompt):

```
$ # ensure serial port is in correct mode
$ stty -F /dev/ttyUSB0 raw speed 115200 -crtscts cs8 -parenb -cstopb -ixon
$ # Put in bootloader mode (device node is /dev/ttyUSB0)
$ echo -n 'R' > /dev/ttyUSB0
$ # flash firmware (inav_6.0.0_MATEKF405.hex)
$ stm32flash -w inav_6.0.0_MATEKF405.hex -v -e 0 -g 0x0 -b 115200 /dev/ttyUSB0
 STM32FLASH: stm32flash -w inav_6.0.0_MATEKF405.hex -v -e 0 -g 0x0 -b 115200 /dev/ttyUSB0
stm32flash 0.7

http://stm32flash.sourceforge.net/

Using Parser : Intel HEX
Location     : 0x8000000
Size         : 554256
Interface serial_posix: 115200 8E1
Version      : 0x31
Option 1     : 0x00
Option 2     : 0x00
Device ID    : 0x0413 (STM32F40xxx/41xxx)
- RAM        : Up to 128KiB  (12288b reserved by bootloader)
- Flash      : Up to 1024KiB (size first sector: 1x16384)
- Option RAM : 16b
- System RAM : 30KiB
Write to memory
Erasing memory
Wrote and verified address 0x08087510 (100.00%) Done.

Starting execution at address 0x08000000... done.

$
```
Above from [mwptools' `fcflash` script](https://stronnag.github.io/mwptools/mwp-miscellaneous-tools/#flashsh-fcflash) which automates the whole thing.

**Reply by alccrl** [reply]
Thanks,
 but my target is to also use configurator by uart, non only to flash the fc.
I will put a link to a video that does this for betaflight, using a free uart.

**Reply by alccrl** [reply]
https://youtu.be/FxSBaoX4v0k

**Reply by stronnag** [reply]
> Thanks, but my target is to also use configurator by uart, non only to flash the fc. I will put a link to a video that does this for betaflight, using a free uart.

Sure, that works too. Just select `/dev/ttyUSB0` in the configurator. MSP on UART1 should be enabled by default

**Comment by alccrl** [comment]
Sadly ...
It definitely does not work ....

I tried again uart1, it says connection succesfully open - 10s -  connection succesfully closed, nothing else.
Same behaviour if I configure with inav-configurator free uart 5 for Msp (using legacy usb connection that is not completely died, up to now) and soldering tx5 and rx 5 instead of t1 rx1.

I do not think that FTDI is faulty, iI just bought it, and leds work properly, it connects to pc either linux or win10.

It could be a problem on fc (f722 std) , either hw (uart/msp/vcp conns)  or sw (does inav 5.1 provide parachute msp by uart ? - not found any info about this online - apart Stronnag response now).
Serial.md doc is quite slim and pragmatic.

There are plenty of tests that I could carry on, but since my quad is more or less already configured and does not output warnings or errors in telemetry or  osd, I will not risk to damage the fc completely while flashing up  and down or soldering up and down.

If more help comes, I will keep it for further investigation , of course.

Thanks,

Alfredo

**Reply by alccrl** [reply]
@stronnag 
Since I tried what You wrote, but it did not work, and You closed the related created issue, I studied a bit the setup.
How do You power F4 when You flash or configure it ?
Which is the hw setup ?
Do You disable vcp before using uart1 as .msp ?
I have a usb to rs232 cable that implies fc is self powered some way (Dce/dte connection), because generally I use tx,rx and gnd on rs232.

I read in forums that when playing in such situations, usb should not be used for powering fc ...

Alfredo

**Reply by stronnag** [reply]
> @stronnag Since I tried what You wrote, but it did not work, and You closed the related created issue, I studied a bit the setup. How do You power F4 when You flash or configure it ? 

5V pad on the FC, from a USB-TTL adapter 

>Which is the hw setup ? 

MATEKF405. MSP on UART 1

>Do You disable vcp before using uart1 as .msp ? 

No.

> I have a usb to rs232 cable that implies fc is self powered some way (Dce/dte connection), because generally I use tx,rx and gnd on rs232.

You should not be using RS232, the board requires TTL.

> I read in forums that when playing in such situations, usb should not be used for powering fc ...

Your choice. How you power it is irrelevant, at least for configuration.  You should not  power by USB to flash via UART (e.g. https://www.st.com/resource/en/application_note/cd00167594-stm32-microcontroller-system-memory-boot-mode-stmicroelectronics.pdf  p 197 et al.)

**Reply by alccrl** [reply]
OK ... thanks !!
..let me summarize:
1)first two rows in configurator enabled (vcp and msp on uart1)
2)pc -> usbcable->ftdi (5v setup)->(rx1 tx1, gnd,vcc).

When I tried ..., this has been the result:
- FTDI led flashes (seems to be in some kind of polling state) but ...
- Configurator does not connect
- STM32CubeProgrammer does not connect (any mode, Matek faqs indicate UART mode).

Do I miss something ...?

Notes:
- bad usb connector on fc is still there, do I have to cut it away ...?
- tried also ftdi with 3.3v, same results
- tried rs232 direct to tx1 rx1, same result.
- I noticed that a minimOSD connected succesfully by FTDI to pc only using 6 wires, not 5 or 4, I ask myself if MSP mode needs some control wire (CTS/DTS or the like). This is not directly related to my problem, but I noticed that.
- On BTFL it seems to be necessary to flash a "rescue" firmware on fc to enable msp on uart1, so it might be that inav configurator does not read exact configuration for Matek F7 (dont know ...) or official firmware for f7 inav is compiled with no uart1 msp option in, in spite of what is declared.

Alfredo
